### ISSUE TITLE
validate doesnt work

### ISSUE THREAD

**Issue by p0fi** [kherge-php/json#5]
I would like to validate a JSON using the validate method.

This is the file for my schema:

```
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "additionalProperties": false,
    "definitions": {},
    "id": "http://example.com/example.json",
    "properties": {
        "commit": {
            "additionalProperties": false,
            "properties": {
                "author_email": {
                    "type": "string"
                },
                "author_name": {
                    "type": "string"
                },
                "authored_date": {
                    "type": "string"
                },
                "committed_date": {
                    "type": "string"
                },
                "committer_email": {
                    "type": "string"
                },
                "committer_name": {
                    "type": "string"
                },
                "id": {
                    "type": "string"
                },
                "message": {
                    "type": "string"
                },
                "parent_ids": {
                    "additionalItems": false,
                    "items": {
                        "type": "string"
                    },
                    "type": "array"
                }
            },
            "type": "object"
        },
        "message": {
            "type": "string"
        },
        "name": {
            "type": "string"
        },
        "release": {
            "additionalProperties": false,
            "properties": {
                "description": {
                    "type": "string"
                },
                "tag_name": {
                    "type": "string"
                }
            },
            "type": "object"
        }
    },
    "type": "object"
}
```

And this is the JSON im trying to validate:

```
{
  "release": {
    "tag_name": "test",
    "description": "some random text here"
  },
  "message": "hallo",
  "commit": {
    "author_email": "random@mail.com",
    "id": "182391378ajskdjj2ß001jklakmsd92",
    "parent_ids": [
      "aksjld9209302aklls002kads"
    ],
    "message": "random string",
    "authored_date": "2016-02-25T15:19:57.000+00:00",
    "author_name": "Pete",
    "committed_date": "2014-02-25T15:19:57.000+00:00",
    "committer_name": "Pete",
    "committer_email": "random@mail.com"
  },
  "name": "test"
}
```

For some reason the validate method never throws anything if I change the JSON on purpose ans is always `null`

This is how im using the method:
```
$schema = $json->decodeFile('valid_json_schema.json');
$json->validate( $schema, $tagData );
```

Am I missing something or is there really something broken?
Regards!

**Comment by kherge** [comment]
Is `$tagData` decoded also, or is it a string containing JSON data?

The `validate()` method expects `$schema` to be a decoded JSON schema and `$tagData` to be decoded JSON data.

**Comment by rjarnutowski** [comment]
I am also experiencing that it doesnt work and i verified agaisnt http://www.jsonschemavalidator.net/

schema
`{
  "type": [
    "object"
  ],
  "properties": {
    "id": {
      "type": "integer"
    },
    "slug": {
      "type": "string"
    },
    "label": {
      "type": "object"
    },
    "position": {
      "type": "integer"
    },
    "region": {
      "enum": [
        "left-col",
        "right-col",
        "main"
      ]
    },
    "row": {
      "type": "integer"
    },
    "settings": {
      "type": "object",
      "oneOf": [
        {"$ref": "#/definitions/requiredOnly"},
        {"$ref": "#/definitions/customText"},
        {"$ref": "#/definitions/image"},
        {"$ref": "#/definitions/html"}
      ]
    },
    "element_definition_slug": {
      "enum": [
        "custom_text",
        "email_address",
        "phone_number",
        "honorific_prefix",
        "given_name",
        "family_name",
        "additional_name",
        "honorific_suffix",
        "postal_code",
        "region",
        "address_line_1",
        "address_ine_2",
        "locality",
        "country",
        "occupation",
        "employer",
        "gender",
        "subscribe",
        "image",
        "static_text",
        "submit",
        "html"
      ]
    }
  },
  "required":[
    "element_definition_slug",
    "slug",
    "label",
    "position",
    "region",
    "row",
    "settings"
  ],
  "definitions": {
    "requiredOnly": {
      "properties": {
        "required": {"type": "boolean"}
      },
      "required": [
        "required"
      ]
    },
    "customText": {
      "properties": {
        "lines": {"type": "integer"},
        "validation": {
          "type": "object",
          "properties": {
            "length": {
              "type": "object",
              "properties": {
                "min": {"type": "integer"},
                "max": {"type": "integer"}
              }
            },
            "text": {
              "type": "object",
              "properties": {
                "pattern": {"type": "string"}
              }
            }
          }
        },
        "required": {"type": "boolean"}
      },
      "required": [
        "lines",
        "validation",
        "required"
      ]
    },
    "image": {
      "properties": {
        "image": {
          "type": "object",
          "properties": {
            "src": {"type": "string"},
            "name": {"type": "string"},
            "alt": {"type": "string"},
            "_links": {
              "type": "object",
              "properties": {
                "asset": {
                  "type": "array",
                  "properties": {
                    "href": {"type": "string"}
                  }
                }
              }
            }
          }
        },
        "link": {
          "type": "object",
          "properties": {
            "href": {"type": "string"},
            "target": {"type": "string"}
          }
        },
        "required": {"type": "boolean"}
      },
      "required": [
        "image",
        "link",
        "required"
      ]
    },
    "html": {
      "type": "object",
      "properties": {
        "html": {"type": "string"}
      },
      "required": [
        "html"
      ]
    }
  }
}
`
This should fail and it passes
`{"id":1,"label":{"source":"custom","en_US":"Element 1"},"slug":"elem-1","row":44,"position":5,"settings":{"invalid":"setting"},"region":"left-col","element_definition_slug":"custom_text"}`

This should pass and it does 
`{"id":1,"label":{"source":"custom","en_US":"Element 1"},"slug":"elem-1","row":44,"position":5,"settings":{"required":true},"region":"left-col","element_definition_slug":"custom_text"}`

**Comment by kherge** [comment]
@p0fi Please give dev-master a try. There was a bug with how error checking was handled and it may be related to your issue. If dev-master works for you, I'll release an update.

@rjarnutowski I've looked at your issue and it appears to be a different issue. I'll continue with yours in the other issue you opened.

**Comment by rjarnutowski** [comment]
@kherge  I found my the issue. I was passing in the decoded value as an array instead of an object. and it will only validate the required properties if it is an object.

**Comment by p0fi** [comment]
Hi @kherge I cannot find any branch called `dev-master` there is only `master` but I assume you mean this one. I tried and didn't think it's working at first. I further investigated and it seems that my understanding of the JSON schema validation was wrong in the first place. What I always did to check if its working is to remove one property such as "release" from my JSON data and tried to validate against the schema. I expected it to fail but it never did. But it would fail if I changed the content of "release" from an object to a string. So what I really needed to do is to update my schema with the following parameter:

`"required": [ "commit", "release", "name", "message" ]` at the top level.

Like this the validation works like I expected it to do. So I don't know if your fixes actually helped in my case. From looking at the changes I think that it should not behave very different but its certainly better to read. An added test case is also welcome though.
